Add Slicer_ASLtoolkit extension

acsenrafilho opened the pull request:
<!--
Thank you for contributing to 3D Slicer!
- To add a new extension with this pull request: Please keep content of "New extension" section and put an 'x' in the brackets for each todo item to indicate that you have accomplished that prerequisite.
- To update an existing extension with this pull request: Please delete all text in this template and just describe which extension is updated and optionally tell us in a sentence what has been changed. To make extension updates easier in the future you may consider replacing specific git hash in your json file by a branch name (for example: `main` for Slicer Preview Releases; `(majorVersion).(minorVersion)` such as `5.6` for Slicer Stable Releases).
-->

# New extension

<!-- To make sure users can find your extension, understand what it is intended for and how to use it, please complete the checklist below. You do not need to complete all the item by the time you submit the pull request, but most likely the changes will only be merged if all the tasks are done. See more information about the submission process here: https://slicer.readthedocs.io/en/latest/developer_guide/extensions.html -->

- [x] Extension has a reasonable name (not too general, not too narrow, suggests what the extension is for)
- [x] Repository name is Slicer+ExtensionName
- [x] Repository is associated with `3d-slicer-extension` GitHub topic so that it is listed [here](https://github.com/topics/3d-slicer-extension). To edit topics, click the settings icon in the right side of "About" section header and enter `3d-slicer-extension` in "Topics" and click "Save changes". To learn more about topics, read https://help.github.com/en/articles/about-topics
- [x] Extension description summarizes in 1-2 sentences what the extension is usable (should be understandable for non-experts)
- [x] Any known related patents must be mentioned in the extension description.
- [x] LICENSE.txt is present in the repository root. MIT  (https://choosealicense.com/licenses/mit/) or Apache (https://choosealicense.com/licenses/apache-2.0/) license is recommended. If source code license is more restrictive for users than MIT, BSD, Apache, or 3D Slicer license then the name of the used license must be mentioned in the extension description.
- [x] Extension URL and revision (scmurl, scmrevision) is correct, consider using a branch name (main, release, ...) instead of a specific git hash to avoid re-submitting pull request whenever the extension is updated
- [x] Extension icon URL is correct (do not use the icon's webpage but the raw data download URL that you get from the download button - it should look something like this: https://raw.githubusercontent.com/user/repo/main/SomeIcon.png)
- [x] Screenshot URLs (screenshoturls) are correct, contains at least one
- [x] Homepage URL points to valid webpage containing the following:
  - [x] Extension name
  - [x] Short description: 1-2 sentences, which summarizes what the extension is usable for
  - [x] At least one nice, informative image, that illustrates what the extension can do. It may be a screenshot.
  - [x] Description of contained modules: at one sentence for each module
  - [x] Tutorial: step-by-step description of at least the most typical use case, include a few screenshots, provide download links to sample input data set
  - [x] Publication: link to publication and/or to PubMed reference (if available)
  - [x] License: We suggest you use a permissive license that includes patent and contribution clauses.  This will help protect developers and ensure the code remains freely available.  We suggest you use the [Slicer License](https://github.com/Slicer/Slicer/blob/main/License.txt) or the [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0). Always mention in your README file the license you have chosen.  If you choose a different license, explain why to the extension maintainers. Depending on the license we may not be able to host your work. Read [here](https://opensource.guide/legal/#which-open-source-license-is-appropriate-for-my-project) to learn more about licenses.
  - [x] Content of submitted json file is consistent with the top-level CMakeLists.txt file in the repository (dependencies, etc. are the same)
- Hide unused features in the repository to reduce noise/irrelevant information:
  - [x] Click `Settings` and in repository settings uncheck `Wiki`, `Projects`, and `Discussions` (if they are currently not used)
  - [x] Click the settings icon next to `About` in the top-right corner of the repository main page and uncheck `Releases` and `Packages` (if they are currently not used)
<!-- If you have any questions or comments then please describe them here. -->

acsenrafilho commented:
@lassoan could you please evaluate this PR as well?

lassoan commented:
There is no CLI modules, so you cannot ask the user about installing a Python package. It is OK to install without asking, like this:

```
try:
    from asltk.asldata import ASLData
    from asltk.reconstruction import CBFMapping
    from asltk.utils import load_image, save_image
except ModuleNotFoundError:
    import slicer.util
    slicer.util.pip_install("asltk")
    from asltk.asldata import ASLData
    from asltk.reconstruction import CBFMapping
    from asltk.utils import load_image, save_image
```

I could not try to actually use any of the modules, because I could not find any tutorial with testing data.

acsenrafilho commented:
Great, @lassoan 
Thanks for the advice. I did not know that, for extensions without CLI, it is ok to install the python package directly.
I have changed the code lines in the main branch
[Link to ASL sample](https://drive.google.com/file/d/1agtHY9SLvC9975L6H0RL6PfIHATnJKcI/view?usp=sharing) 

BTW, I have also added a small data test zip file here to assist with a (brief) testing.
I will also add this example test data to the documentation, which will be publicly available to users and developers.

acsenrafilho commented:
@lassoan , just to say that I have updated the extension documentation to give a quick dataset tutorial to assist for further testing.
See more details in the section `Quick tutorial` in the [documentation](https://slicer-asltoolkit.readthedocs.io/)

I already included the  `How to install` section assuming that (soon) this extension will be live at Slicer Extension Manager (finger crossed 🤞 )

I hope that this will be useful for the general community. Any thoughts are very welcome!

BTW, thank you very much for all the help!

lassoan commented:
Thank you for your patience. Have you tried if `slicer.util.messageBox` actually works in Python CLI modules? I would expect it to fail. It is because that GUI CLI modules are not meant to provide any interactivity or GUI (other than the GUI that is automatically generated from the .xml file). If there is any message that you want to display to the user then print it to the standard output or standard error. If there is an error then exit with a non-zero exit code.

acsenrafilho commented:
Hi @lassoan

Thank you for the suggestion. I made the change in the print messages, using the direct standard output. All the `slicer.util.messagebox` and similar methods were replaced to be print out as usual. 

Is there anything else to be adjusted? 

Thank you again for all the tips!

lassoan commented:
It should be good, only the extension name and category needs some attention.

❌ Check extension name failed: extension name should not start with 'Slicer'. Please, consider changing it to '_ASLtoolkit'.

Could you update the project name in the top-level CMakeLists.txt file to be `ASLToolkit` (or maybe `ASLtoolkit`)? It is good to have the `Slicer` prefix in the repository name, but not in the extension's name.

❌ Check category failed: Category Arterial Sping Labelling is unknown. Consider using any of the known extensions instead: Active Learning, Analysis, Auto3dgm, BigImage, Cardiac, Chest Imaging Platform, Conda, Converters, DICOM, DSCI, Developer Tools, Diffusion, Examples, Exporter, FTV Segmentation, Filtering, Filtering.Morphology, Filtering.Vesselness, Holographic Display, IGT, Informatics, Netstim, Neuroimaging, Nuclear Medicine, Orthodontics, Osteotomy Planning, Otolaryngology, Photogrammetry, Pipelines, Planning, Printing, Quantification, Radiotherapy, Registration, Remote, Rendering, SPHARM, Segmentation, Sequences, Shape Analysis, Shape Regression, Shape Visualization, Simulation, SlicerCMF, SlicerMorph, Spectral Imaging, Supervisely, Surface Models, SurfaceLearner, Tomographic Reconstruction, Tracking, Tractography, Training, Ultrasound, Utilities, Vascular Modeling Toolkit, Virtual Reality, VisSimTools, Web System Tools, Wizards

Would any of these existing categories work for your extension?
